src/Storages/MergeTree/GinIndexStore.cpp

```diff
@@ -35,123 +35,123 @@ const CompressionCodecPtr & GinIndexCompressionFactory::zstdCodec()
 
-        writeVarUInt(size, buffer);
-        written_bytes += getLengthOfVarUInt(size);
+        header = (header << 2) | (ROARING_COMPRESSED_MASK << 1) | ROARING_CONTAINER_MASK;
 
+        writeVarUInt(header, buffer);
         writeVarUInt(compressed_size, buffer);
-        written_bytes += getLengthOfVarUInt(compressed_size);
-
         buffer.write(memory.data(), compressed_size);
-        written_bytes += compressed_size;
+
+        return getLengthOfVarUInt(header) + getLengthOfVarUInt(compressed_size) + compressed_size;
     }
     else
     {
-        for (size_t i = 0; i <  rowid_lst_length; ++i)
-        {
-            writeVarUInt(rowid_lst[i], buffer);
-            written_bytes += getLengthOfVarUInt(rowid_lst[i]);
-        }
-    }
+        header = (header << 2) | (ROARING_UNCOMPRESSED_MASK << 1) | ROARING_CONTAINER_MASK;
+
+        writeVarUInt(header, buffer);
+        buffer.write(buf.data(), uncompressed_size);
 
-    return written_bytes;
+        return getLengthOfVarUInt(header) + uncompressed_size;
+    }
 }
 
 GinIndexPostingsListPtr GinIndexPostingsBuilder::deserialize(ReadBuffer & buffer)
 {
-    UInt8 postings_list_size = 0;
-    buffer.readStrict(reinterpret_cast<char &>(postings_list_size));
+    /**
+     * Header value maps into following states:
+     * The lowest bit indicates if values are stored as an array or Roaring bitmap
+     * In case of array container, the rest of the bits is the number of entries in the array.
+     * In case of Roaring bitmap, the second lowest bit indicates if Roaring bitmap is compressed or uncompressed, the rest of the bits is the uncompressed size.
+     */
+    UInt64 header = 0;
+    readVarUInt(header, buffer);
 
-    if (postings_list_size == USES_BIT_MAP)
+    if (header & ARRAY_CONTAINER_MASK) /// Array
     {
-        size_t size = 0;
-        size_t compressed_size = 0;
-        readVarUInt(size, buffer);
-        readVarUInt(compressed_size, buffer);
-        auto buf = std::make_unique<char[]>(compressed_size);
-        buffer.readStrict(reinterpret_cast<char *>(buf.get()), compressed_size);
-
-        Memory<> memory;
-        memory.resize(size);
-        const auto & codec = GinIndexCompressionFactory::zstdCodec();
-        codec->decompress(buf.get(), static_cast<UInt32>(compressed_size), memory.data());
-
-        GinIndexPostingsListPtr postings_list = std::make_shared<GinIndexPostingsList>(GinIndexPostingsList::read(memory.data()));
+        UInt64 num_entries = (header >> 1);
+        std::vector<UInt32> values(num_entries);
+        for (size_t i = 0; i < num_entries; ++i)
+            readVarUInt(values[i], buffer);
 
+        GinIndexPostingsListPtr postings_list = std::make_shared<GinIndexPostingsList>();
+        postings_list->addMany(values.size(), values.data());
         return postings_list;
     }
+    else /// Roaring
+    {
+        header >>= 1;
```

Comment:
For testing, I added `chassert(header & ROARING_CONTAINER_MASK);` _before_ l. 129, then ran all text index tests using a debug build and it crashed :-( What was I doing wrong here?

Reply:
Add the assert, then run

```sql
SET allow_experimental_full_text_index = 1;

CREATE TABLE tab(k UInt64, s String, INDEX af(s) TYPE text(tokenizer = 'ngram', ngram_size = 2) GRANULARITY 1)
            ENGINE = MergeTree() ORDER BY k
            SETTINGS index_granularity = 2, index_granularity_bytes = '10Mi';

INSERT INTO tab VALUES (101, 'Alick a01'), (102, 'Blick a02'), (103, 'Click a03'), (104, 'Dlick a04'), (105, 'Elick a05'), (106, 'Alick a06'), (107, 'Blick a07'), (108, 'Click a08'), (109, 'Dlick a09'), (110, 'Elick a10'), (111, 'Alick b01'), (112, 'Blick b02'), (113, 'Click b03'), (114, 'Dlick b04'), (115, 'Elick b05'), (116, 'Alick b06'), (117, 'Blick b07'), (118, 'Click b08'), (119, 'Dlick b09'), (120, 'Elick b10');

SELECT * FROM tab WHERE s == 'Alick a01';
```

Reply:
Oh, didn't read your comment

Reply:
No, then it is fine of course :)

Reply:
that's fine :) I also added the same assert then also surprised for a moment :)